pkg/controller/cluster/cluster_controller.go

```diff
@@ -35,114 +40,141 @@ import (
 	// Create a new controller
-	c, err := controller.New("cluster-controller", mgr, controller.Options{Reconciler: r})
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	// Watch for changes to Cluster
 	err = c.Watch(&source.Kind{Type: &clusterv1.Cluster{}}, &handler.EnqueueRequestForObject{})
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	return nil
+	return c, nil
 }
 
 var _ reconcile.Reconciler = &ReconcileCluster{}
 
 // ReconcileCluster reconciles a Cluster object
 type ReconcileCluster struct {
 	client.Client
-	scheme   *runtime.Scheme
-	actuator Actuator
+	scheme           *runtime.Scheme
+	controller       controller.Controller
+	recorder         record.EventRecorder
+	externalWatchers sync.Map
 }
 
-func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+func (r *ReconcileCluster) Reconcile(request reconcile.Request) (_ reconcile.Result, reterr error) {
+	ctx := context.TODO()
+
+	// Fetch the Cluster instance.
 	cluster := &clusterv1.Cluster{}
-	err := r.Get(context.Background(), request.NamespacedName, cluster)
+	err := r.Get(ctx, request.NamespacedName, cluster)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
 			// Object not found, return.  Created objects are automatically garbage collected.
 			// For additional cleanup logic use finalizers.
 			return reconcile.Result{}, nil
 		}
+
 		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
 	}
 
-	name := cluster.Name
-	klog.Infof("Running reconcile Cluster for %q", name)
+	// Store Cluster early state to allow patching.
+	patchCluster := client.MergeFrom(cluster.DeepCopy())
 
-	// If object hasn't been deleted and doesn't have a finalizer, add one
-	// Add a finalizer to newly created objects.
-	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
-		finalizerCount := len(cluster.Finalizers)
-
-		if !util.Contains(cluster.Finalizers, metav1.FinalizerDeleteDependents) {
-			cluster.Finalizers = append(cluster.ObjectMeta.Finalizers, metav1.FinalizerDeleteDependents)
+	// Always issue a Patch for the Cluster object and its status after each reconciliation.
+	defer func() {
+		gvk := cluster.GroupVersionKind()
+		if err := r.Client.Patch(ctx, cluster, patchCluster); err != nil {
+			klog.Errorf("Error Patching Cluster %q in namespace %q: %v", cluster.Name, cluster.Namespace, err)
+			if reterr != nil {
+				reterr = err
+			}
+		}
+		// TODO(vincepri): This is a hack because after a Patch, the object loses TypeMeta information.
+		// Remove when https://github.com/kubernetes-sigs/controller-runtime/issues/526 is fixed.
+		cluster.SetGroupVersionKind(gvk)
+		if err := r.Client.Status().Patch(ctx, cluster, patchCluster); err != nil {
+			klog.Errorf("Error Patching Cluster status %q in namespace %q: %v", cluster.Name, cluster.Namespace, err)
+			if reterr != nil {
```

Comment:
If you return early if the main patch failed, then if we made it this far, reterr should still be nil, right?